### PR TITLE
build(drive-abci): debug docker image with grovedb visualizer and tokio console

### PR DESCRIPTION
**Pull request by lklimek**
## Issue being fixed or feature implemented

When building drive-abci image, we don't want to include tools like grovedb visualizer or tokio console.
However, for some use cases, we need these features.

We need to build additional debug package that will accompany normal release and contain forementioned features.

## What was done?

1. Refactored docker image tag detection algorithm. Right now, only full releases will generate tags like MAJOR, MAJOR.MINOR, latest, etc.
2. Added manual workflow checkbox "Only build Drive image"
3. Fixed bug that could lead to use of wrong build of app retrieved from cache.
4. Added drive-abci debug image build that will have docker tag suffixed with `-debug`, for example for `1.2.3-dev.1`,  we build `dashpay/drive-abci:1.2.3-dev.1` and `dashpay/drive-abci:1.2.3-dev.1-debug`.
5. By default, `tokio_unstable`, `tokio console` and `grovedbg` are enabled. For release builds (release cargo profile), this is changed on Dockerfile level - drive-abci builds with `--no-default-features`, and we use separate config file that has `tokio_unstable` disabled.
6. Normal images are built in `release` cargo profile (as previously), `-debug` one in `dev` cargo profile.
7. In drive-abci main.rs, added info log containing version info and list of enabled features, for easier debugging.

### Manual workflow run

https://github.com/dashpay/platform/actions/runs/10508658962

###  Prerelease v1.1.0-pr.2012.6

https://github.com/dashpay/platform/actions/runs/10508221419

#### Debug

```bash
docker run --rm -ti dashpay/drive:1.1.0-pr.2012.6-debug status
 2024-08-22T13:02:08.132339Z  INFO drive_abci: drive-abci server initializing, version: "1.1.0-dev.1", features: "console,grovedbg,mocks", rust: "1.76"
```
#### Release

```bash
docker run --rm -ti dashpay/drive:1.1.0-pr.2012.6 status
2024-08-22T13:03:19.811619Z  INFO drive_abci: drive-abci server initializing, version: "1.1.0-dev.1", features: "mocks", rust: "1.76"
```

### Not tested

* full release

## Breaking Changes

We no longer tag images with tags like 1-dev, latest-dev, etc.

## Checklist:

- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added or updated relevant unit/integration/functional/e2e tests
- [ ] I have added "!" to the title and described breaking changes in the corresponding section if my code contains any
- [ ] I have made corresponding changes to the documentation if needed

**For repository code-owners and collaborators only**
- [ ] I have assigned this pull request to a milestone

### CODE REVIEW

**Comment by QuantumExplorer** [comment]
do not merge for now

File: .github/workflows/release-docker-image.yml [hunk 1]
```diff
@@ -52,10 +60,10 @@ jobs:
         with:
           image_name: ${{ inputs.image_name }}
           image_org: ${{ inputs.image_org }}
-          image_version: ${{ github.event_name == 'workflow_dispatch' && github.event.inputs.tag || github.event.release.tag_name }}
+          image_version: ${{ inputs.tag || github.event.release.tag_name }}
```
Comment (shumkov) [inline .github/workflows/release-docker-image.yml:63]: we have default value for `inputs.tag` so `||` will never happen

File: .github/workflows/release-docker-image.yml [hunk 2]
```diff
@@ -98,11 +106,7 @@ jobs:
         with:
           result-encoding: string
           script: |
-            return (
-              context.eventName === 'workflow_dispatch'
-                ? '${{ github.event.inputs.tag }}'
-                : context.payload.release.tag_name
-            );
+            return ( "${{ inputs.tag }}" ||  context.payload.release.tag_name );
```
Comment (shumkov) [inline .github/workflows/release-docker-image.yml:109]: why do we still need this if `inputs.tag` default is `github.event.inputs.tag || github.event.release.tag_name`?

File: .github/workflows/release-docker-image.yml [hunk 3]
```diff
@@ -125,20 +128,17 @@ jobs:
...
-            type=match,pattern=v(\d+),group=1,value=${{ steps.version.outputs.result }}
-            type=match,pattern=v(\d+.\d+),group=1,value=${{ steps.version.outputs.result }}
-            type=match,pattern=v(\d+.\d+.\d+),group=1,value=${{ steps.version.outputs.result }}
-            type=match,pattern=v(.*),group=1,value=${{ steps.version.outputs.result }},suffix=
+            type=semver,pattern={{version}},priority=950,value=${{ steps.version.outputs.result }},enable=${{ steps.version.outputs.result != '' }}
+            type=semver,pattern={{major}}.{{minor}},value=${{ steps.version.outputs.result }},enable=${{ steps.version.outputs.result != '' }}
+            type=semver,pattern={{major}},value=${{ steps.version.outputs.result }},enable=${{ steps.version.outputs.result != '' }}
+            type=ref,event=pr
           flavor: |
-            suffix=${{ steps.suffix.outputs.result }},onlatest=true
```
Comment (shumkov) [inline .github/workflows/release-docker-image.yml:133]: why did we removed suffix?

File: .github/workflows/release.yml [hunk 4]
```diff
@@ -9,6 +9,10 @@ on:
       tag:
         description: "Version (i.e. v0.22.3-pre.2)"
         required: true
+      only_drive:
```
Comment (shumkov) [inline .github/workflows/release.yml:12]: this is for test only?

File: .github/workflows/release.yml [hunk 5]
```diff
@@ -136,40 +145,60 @@ jobs:
       image_org: dashpay
       image_name: drive
       target: drive-abci
+      tag: ${{ inputs.tag || github.event.release.tag_name }}
```
Comment (shumkov) [inline .github/workflows/release.yml:148]: didn't we miss `github.event.inputs.tag` ?

File: packages/rs-drive-abci/Cargo.toml [hunk 6]
```diff
@@ -101,8 +101,8 @@ rocksdb = { version = "0.22.0" }
 integer-encoding = { version = "4.0.0" }
 
 [features]
-# TODO: Remove console for production
-default = ["mocks", "console"]
+
+default = ["mocks", "console", "grovedbg"]
```
Comment (shumkov) [inline packages/rs-drive-abci/Cargo.toml:105]: I think the idea to remove it from default

**Review by shumkov — CHANGES_REQUESTED** [review]

File: packages/rs-drive-abci/Cargo.toml (hunk 6 above)
Comment (lklimek) [reply packages/rs-drive-abci/Cargo.toml]: ok

File: .github/workflows/release-docker-image.yml (hunk 3 above)
Comment (lklimek) [reply .github/workflows/release-docker-image.yml:133]: reverted

File: .cargo/config-release.toml [hunk 7]
```diff
@@ -0,0 +1,11 @@
+# This is a config file used by Docker build process when building release version of drive-abci.
+# Hopefully it can be removed once profile-rustflags is stabilized, see https://github.com/rust-lang/cargo/issues/10271
+
+[target.aarch64-unknown-linux-musl]
```
Comment (shumkov) [inline .cargo/config-release.toml:4]: I think you can do something like that:

```
[target.aarch64-unknown-linux-musl]
rustflags = ["-C", "target-feature=-crt-static"]

[target.aarch64-unknown-linux-musl.profile.dev]
rustflags = ["-C", "target-feature=-crt-static", "--cfg", "tokio_unstable"]

[target.x86_64-unknown-linux-musl]
rustflags = ["-C", "target-feature=-crt-static"]

[target.x86_64-unknown-linux-musl.profile.dev]
rustflags = ["-C", "target-feature=-crt-static", "--cfg", "tokio_unstable"]

[target.aarch64-unknown-linux-gnu]
linker = "aarch64-linux-gnu-gcc"

[build.profile.dev]
rustflags = ["--cfg", "tokio_unstable"]
```

**Review by shumkov — CHANGES_REQUESTED** [review]
I don't see where we add f

**Comment by lklimek** [comment]
> I don't see where we add f

If you mean "where we add features", it's in Dockerfile lines 202-209,  https://github.com/dashpay/platform/pull/2012/files#diff-dd2c0eb6ea5cfc6c4bd4eac30934e2d5746747af48fef6da689e85b752f39557R202-R209

File: .cargo/config-release.toml (hunk 7 above)
Comment (lklimek) [reply .cargo/config-release.toml:4]: Unfortunately, it doesn't work on `stable`.

Comment (shumkov) [reply .cargo/config-release.toml:4]: Bump Rust version? )))

**Review by shumkov — APPROVED** [review]

File: .cargo/config-release.toml (hunk 7 above)
Comment (lklimek) [reply .cargo/config-release.toml:4]: error: expected a string, but found a table for `dev`

**Review by QuantumExplorer — APPROVED** [review]

**Pull request merged**